hummingbot/connector/exchange/bittrex/bittrex_api_user_stream_data_source.py

```diff
@@ -1,160 +1,98 @@
-
-        output: Dict[str, Any] = {"event_type": None, "content": None, "error": None}
-        msg: Dict[str, Any] = ujson.loads(msg)
-
-        if _is_auth_notification(msg):
-            output["event_type"] = "re-authenticate"
-
-        elif _is_heartbeat(msg):
-            output["event_type"] = "heartbeat"
-
-        elif _is_balance_delta(msg) or _is_order_delta(msg) or _is_execution_event(msg):
-            output["event_type"] = msg["M"][0]["M"]
-            output["content"] = _decode_message(msg["M"][0]["A"][0])
-
-        return output
-
-    async def listen_for_user_stream(self, output: asyncio.Queue):
-        while True:
-            try:
-                self._websocket_connection = signalr_aio.Connection(BITTREX_WS_FEED, session=None)
-                self.hub = self._websocket_connection.register_hub("c3")
-
-                await self.authenticate()
-                self.hub.server.invoke("Subscribe", ["heartbeat", "order", "balance", "execution"])
-                self._websocket_connection.start()
-
-                async for raw_message in self._socket_user_stream(self._websocket_connection):
-                    decode: Dict[str, Any] = self._transform_raw_message(raw_message)
-                    self.logger().debug(f"Got ws message {decode}")
-                    if decode.get("error") is not None:
-                        self.logger().error(decode["error"])
-                        continue
-
-                    content_type = decode.get("event_type")
-                    if content_type is not None:
-                        if content_type in ["balance", "order", "execution"]:
-                            output.put_nowait(decode)
-                        elif content_type == "re-authenticate":
-                            await self.authenticate()
-                        elif content_type == "heartbeat":
-                            self.logger().debug("WS heartbeat")
-                            continue
-
-            except asyncio.CancelledError:
+            channels = ["order", "balance", "execution"]
+            response = await self.invoke('Subscribe', [channels])
+            flag = True
+            for i in range(len(channels)):
+                if response[i]['Success']:
+                    self.logger().info("Subscription to '" + channels[i] + "' successful")
+                else:
+                    flag = False
+                    print("Subscription to '" + channels[i] + "' failed: " + response[i]["ErrorCode"])
+            if not flag:
                 raise
+        except asyncio.CancelledError:
+            raise
+        except Exception:
+            self.logger().error("Failed to subscribe to all private channels")
+            raise
+
+    async def _process_websocket_messages(self, websocket_assistant, queue: asyncio.Queue):
+        """
+        Overriding method for Bittrex signalr
+        """
+        async for raw_message in _socket_stream(conn=websocket_assistant):
```

Comment:
As in the order book data source, we should not break signalr components encapsulation. We should use the events system the framework provides.

Reply:
Reworked the implementation of signalr with inspiration from the CoinDCX implementation.